client/concordclient/include/client/concordclient/trc_queue.hpp

```diff
@@ -0,0 +1,112 @@
+// Concord
+//
+// Copyright (c) 2020-2022 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0
+// License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#pragma once
+
+#include <condition_variable>
+#include <boost/lockfree/spsc_queue.hpp>
+#include <list>
+#include <memory>
+#include <thread>
+#include <mutex>
+#include <exception>
+#include <chrono>
+#include <variant>
+
+#include "assertUtils.hpp"
+#include "event_update.hpp"
+
+namespace concord::client::concordclient {
+
+typedef std::variant<Update, EventGroup> EventVariant;
+
+class TrcQueue {
+ private:
+  boost::lockfree::spsc_queue<EventVariant*> queue_data_;
+  std::exception_ptr exception_;
+  std::mutex mutex_;
+  std::condition_variable condition_;
+  static constexpr size_t kQueueDataSize{10000u};
+
+ public:
+  // Construct a TrcQueue.
+  TrcQueue() : queue_data_(kQueueDataSize) {}
+
+  // Copying or moving a TrcQueue is explicitly disallowed, as we do not know of a compelling use case
+  // requiring copying or moving TrcQueue, we believe semantics for these operations are likely to be
+  // messy in some cases, and we believe implementation may be non-trivial. We may revisit the decision to disallow
+  // these operations should compelling use cases for them be found in the future.
+  TrcQueue(const TrcQueue& other) = delete;
+  TrcQueue(const TrcQueue&& other) = delete;
+  TrcQueue& operator=(const TrcQueue& other) = delete;
+  TrcQueue& operator=(const TrcQueue&& other) = delete;
+
+  void clear() { queue_data_.reset(); }
+
+  void push(EventVariant* update) {
+    queue_data_.push(std::move(update));
+    condition_.notify_one();
+  }
+
+  std::unique_ptr<EventVariant> getElementFromQueue() {
+    if (queue_data_.read_available()) {
+      EventVariant* update = new EventVariant;
+      queue_data_.pop(update);
+      return std::unique_ptr<EventVariant>(update);
+    }
+    if (exception_) {
+      auto e = exception_;
+      exception_ = nullptr;
+      std::rethrow_exception(e);
+    }
+    return nullptr;
+  }
+
+  std::unique_ptr<EventVariant> popTill(std::chrono::milliseconds timeout) {
```

Comment:
Why do we need this method for TRC queue?

Reply:
Because event service uses popTill ([ref](https://github.com/vmware/concord-bft/blob/d78090cd03393b8ed3db4a515c1c5f9d961bb692/client/clientservice/src/event_service.cpp#L84)).